lib/ide/src/controller/graph.rs

```diff
@@ -0,0 +1,177 @@
+//! Graph Controller.
+//!
+//! This controller provides access to a specific graph. It lives under a module controller, as
+//! each graph belongs to some module.
+
+pub mod mock;
+
+use crate::prelude::*;
+
+use flo_stream::Subscriber;
+
+pub use double_representation::graph::Id;
+pub use controller::node::Position;
+pub use controller::notification;
+
+
+
+// ==============
+// === Errors ===
+// ==============
+
+#[derive(Clone,Debug,Fail)]
+#[fail(display = "Node by ID {} was not found.", _0)]
+struct NodeNotFound(ast::ID);
+
+
+
+// ===================
+// === NewNodeInfo ===
+// ===================
+
+/// Describes the node to be added.
+#[derive(Clone,Debug)]
+pub struct NewNodeInfo {
+    /// Expression to be placed on the node
+    pub expression:String,
+    /// Visual node position in the graph scene.
+    pub position:Position,
+    /// ID to be given to the node.
+    pub id:Option<ast::ID>,
+    /// Where line created by adding this node should appear.
+    pub location_hint:LocationHint
+}
+
+/// Describes the desired position of the node's line in the graph's code block.
+#[derive(Clone,Copy,Debug)]
+pub enum LocationHint {
+    /// Try placing this node's line before the line described by id.
+    Before(ast::ID),
+    /// Try placing this node's line after the line described by id.
+    After(ast::ID),
```

Comment:
Some ids are named `ID` and some are named `Id` . The later is correct. pls fix.